### PR TITLE
runtime/interpreter: Check the paths first for faster and cheaper detection

### PR DESCRIPTION
**Pull request by kakkoyun**
Signed-off-by: Kemal Akkoyun <kakkoyun@gmail.com>

https://pprof.me/beb2cec956f42f3b78f38fc2a941ffdf
Ruby: https://pprof.me/4621cebc0f971f2e0f5a8bcfce78a90b
Python: https://pprof.me/14292dfb9a915878533098f53aa9b374

### CODE REVIEW

File: pkg/runtime/ruby/ruby.go [hunk 1]
```diff
@@ -45,39 +44,61 @@ import (
...
+	if err != nil {
+		return false, fmt.Errorf("error reading process maps: %w", err)
 	}
 
-	if !ruby {
-		if ruby, err = runtime.IsSymbolNameInDynamicSymbols(ef, rubyIdentifyingSymbols); err != nil && !errors.Is(err, elf.ErrNoSymbols) {
-			return ruby, fmt.Errorf("search dynamic symbols: %w", err)
+	for _, mapping := range maps {
+		if isRubyLib(mapping.Pathname) {
+			// Let's make sure it's a ruby process by checking the ELF file.
```
Comment (javierhonduco) [inline pkg/runtime/ruby/ruby.go:91]: nit:
```suggestion
			// Let's make sure it's a Ruby process by checking the ELF file.
```

File: pkg/runtime/ruby/ruby.go [hunk 2]
```diff
@@ -45,39 +44,61 @@ import (
...
 		return false, err
 	}
 
-	// Let's make sure it's a python process by checking the ELF file.
-	ef, err := elf.Open(absolutePath(proc, exe))
-	if err != nil {
-		return false, fmt.Errorf("open elf file: %w", err)
-	}
+	if isRubyBin(exe) {
+		// Let's make sure it's a ruby process by checking the ELF file.
```
Comment (javierhonduco) [inline pkg/runtime/ruby/ruby.go:74]: nit: 
```suggestion
		// Let's make sure it's a Ruby process by checking the ELF file.
```

File: pkg/runtime/python/python.go [hunk 3]
```diff
@@ -60,35 +61,50 @@ const (
...
-	if python, err = runtime.IsSymbolNameInSymbols(ef, pythonIdentifyingSymbols); err != nil && !errors.Is(err, elf.ErrNoSymbols) {
-		return python, fmt.Errorf("search symbols: %w", err)
+	// If the executable is not a Python interpreter, let's check the memory mappings.
+	maps, err := proc.ProcMaps()
+	if err != nil {
+		return false, fmt.Errorf("error reading process maps: %w", err)
 	}
+	for _, mapping := range maps {
+		if isPythonLib(mapping.Pathname) {
+			// Let's make sure it's a python process by checking the ELF file.
```
Comment (javierhonduco) [inline pkg/runtime/python/python.go:97]: nit: 
```suggestion
			// Let's make sure it's a Python process by checking the ELF file.
```

File: pkg/runtime/python/python.go [hunk 4]
```diff
@@ -60,35 +61,50 @@ const (
...
+	[]byte(pythonRuntimeSymbol),
+	[]byte(pythonThreadStateSymbol),
+}
+
 func absolutePath(proc procfs.Proc, p string) string {
 	return path.Join("/proc/", fmt.Sprintf("%d", proc.PID), "/root/", p)
 }
 
 func IsInterpreter(proc procfs.Proc) (bool, error) {
+	// First, let's check the executable`pathname since it's the cheapest and fastest.
```
Comment (javierhonduco) [inline pkg/runtime/python/python.go:74]: nit: 
```suggestion
	// First, let's check the executable's pathname since it's the cheapest and fastest.
```

**Review by javierhonduco — APPROVED** [review]
LGTM! Perhaps I missed it but I could not see Python or Ruby stacks in the profile, could you add them if they aren't present?

**Comment by kakkoyun** [comment]
> LGTM! Perhaps I missed it but I could not see Python or Ruby stacks in the profile, could you add them if they aren't present?

The profile just from the agent, not the targets. 

I can add them as well.

**Comment by javierhonduco** [comment]
We don't have integration tests for the interpreters, while I don't think this PR introduces any bugs, I think it would be a good idea to verify that everything still works as expected

**Comment by kakkoyun** [comment]
> We don't have integration tests for the interpreters, while I don't think this PR introduces any bugs, I think it would be a good idea to verify that everything still works as expected

Yes, of course, I verified it. I'll attach a profile.

**Pull request merged**